File: library_management/library_management/page/import_books/import_books.js
frappe.pages['import-books'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'None',
		single_column: true
	});

	// load html content
	$(frappe.render_template("import_books", {})).appendTo(page.body);

	page.book_importer = new BookImporter(page)
}

class BookImporter {
	constructor(page) {
		this.page = page;
		this.bind_events();
	}

	bind_events() {
		$("#import_books_btn").on('click', () => this.import_books());
	}

	import_books() {
		const number_of_books = parseInt($("#number_of_books").val()) || 20;
		const title = $("#title").val() || '';
		const authors = $("#authors").val() || '';
		const isbn = $("#isbn").val() || '';
		const publisher = $("#publisher").val() || '';
		const starting_page = parseInt($("#starting_page").val()) || 20;

		// Validate input
		if (number_of_books <= 0 || number_of_books > 100) {
			frappe.throw(__('Number of books must be between 1 and 100'));
			return;
		}

		// Show progress section
		$('.import-results').show();
		$('#import_status').text('Fetching books from Frappe Library API...');
		$('#import_progress').width('0%').attr('aria-valuenow', 0).text('0%');
		$('#books_list').empty();

		// server call implimentation
		frappe.call({
			method: "library_management.library_management.page.import_books.import_books.import_books",
			args: {
				"number_of_books": number_of_books,
				"title": title,
				"authors": authors,
				"isbn": isbn,
				"publisher": publisher,
				"starting_page": starting_page
			},
			callback: (r) => {
				if (r.message) {
					const result = r.message;

					//updating results
					$('#import_status').text(`Imported ${result.imported_count} books succesfullly. ${result.existing_count} books already existed.`);

					//set progress 100%
					$("#import_progrss").width('100%').attr("aria-valuenow", 100).text('100%');

					//display the list of imported books
					if (result.imported_books.length > 0) {
						let html = '<div class="table-responsive"><table class="table table-bordered">';
						html += '<thead><tr><th>ISBN</th><th>Title</th><th>Authors</th><th>Publisher</th></tr></thead>';
						html += '<tbody>';

						result.imported_books.forEach(book => {
							html += `<tr>
								<td>${book.isbn}</td>
								<td>${book.title}</td>
								<td>${book.authors}</td>
								<td>${book.publisher}</td>
							</tr>`;
						});

						html += '</tbody></table></div>';
						$("#books_list").html(html);
					} else {
						$("#books_list").html('<div class="alert alert-warning "> No new books where imported.</div>');
					}
				}
			},
			error: (r) => {
				$("#import-status").text('Error Importing Books : ' + r.message);
				$("#import-progress").addClass('bg-danger');
			}
		});
	}
}

File: library_management/library_management/page/import_books/import_books.py
import frappe, requests


@frappe.whitelist()
def import_books(
    number_of_books, title="", authors="", isbn="", publisher="", starting_page=1
):
    """
    Import books from Frappe Library API
    Args:
        number_of_books: Number of books to import (max 100)
        title: Filter by book title
        authors: Filter by book authors
        isbn: Filter by book ISBN
        publisher: Filter by book publisher
        starting_page: Starting page number for API pagination

    Returns:
        dict: Result of import operation
    """
    number_of_books = int(number_of_books)
    starting_page = int(starting_page)

    if number_of_books <= 0 or number_of_books > 100:
        frappe.throw("Numbeer of books must be between 1 to 100")

    # calculate how many api we need to make
    books_per_page = 20
    num_pages = int(number_of_books / books_per_page)

    all_books = []
    current_page = starting_page

    for i in range(num_pages + 1):
        params = {"page": current_page}

        if title:
            params["title"] = title

        if authors:
            params["authors"] = authors

        if isbn:
            params["isbn"] = isbn

        if publisher:
            params["publisher"] = publisher

        try:
            response = requests.get(
                "https://frappe.io/api/method/frappe-library", params=params
            )
            response.raise_for_status()

            books_data = response.json().get("message", [])

            if not books_data:
                break

            all_books.extend(books_data)
            current_page += 1

            if len(all_books) >= number_of_books:
                all_books = all_books[:number_of_books]
                break

        except Exception as e:
            frappe.log_error(
                f"Error fetching books from API: {str(e)}", "Library API Import Error"
            )
            frappe.throw(f"Error fetching books: {str(e)}")

    imported_books = []
    existing_count = 0

    # check book is alreagy exist
    for book_data in all_books:
        if frappe.db.exists("Book", {"isbn": book_data.get("isbn")}):
            existing_count += 1
            continue

        # create new books
        try:
           new_book = frappe.new_doc("Book")

           new_book.title = book_data.get("title")
           new_book.author = book_data.get("authors")
           new_book.isbn = book_data.get("isbn")
           new_book.publisher = book_data.get("publisher")
           new_book.num_pages = book_data.get("num_pages")
           new_book.publication_date = book_data.get("publication_date")
           new_book.langague_code = book_data.get("langague_code")
           new_book.avarage_rating = book_data.get("avarage_rating")
           new_book.stock = 1

           new_book.insert()
           frappe.db.commit()

           imported_books.append({
               "isbn"  : new_book.isbn,
               "authors" : new_book.author,
               "title" : new_book.title,
               "publisher" : new_book.publisher
           })

        except Exception as e:
            frappe.log_error(f"Error importing book {book_data.get('title')}: {str(e)}", "Library Book Import Error")

    return {
        "imported_count": len(imported_books),
        "existing_count": existing_count,
        "total_fetched" : len(all_books),
        "imported_books" : imported_books
    }
